File: FunnyAlgorithms-master/sort/heapSort/heapSort.py
from random import randint

#  heap sort

def build_max_heap(arr):
    length = len(arr)
    for i in range(int(length/2), -1, -1):
        max_heapify(arr, i, length)

    #  heap sort

    for i in range(length-1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        max_heapify(arr, 0, i)


def max_heapify(arr, i, heapSize):
    left = (2 * i) + 1
    right = (2 * i) + 2
    if left < heapSize and arr[left] > arr[i]:
        largest = left
    else:
        largest = i

    if right < heapSize and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        max_heapify(arr, largest, heapSize)


array = [randint(1, 100) for _ in range(20)]
print(array)
build_max_heap(array)
print(array)

File: algorithms/sort/heapSort/heapSort.cpp
#include <iostream>

using namespace std;

#define Child(i, pos)  2*i+pos

void Max_Heapify(int Heap[], int i, int Len){

	int Left_Child = Child(i, 1);		//1 to indicate Left Child
	int Right_Child = Child(i, 2);		//2 to indicate Right Child
	int temp, Largest = i;

	//Find the largest element among the Root, Left child and Right child
	if(Left_Child < Len and Heap[Left_Child] > Heap[i]) Largest = Left_Child;
	if(Right_Child < Len and Heap[Right_Child] > Heap[Largest]) Largest = Right_Child;

	//if root is not the largest, Swap root with largest and Heapify again
	if(Largest != i){
		temp = Heap[Largest];
		Heap[Largest] = Heap[i];
		Heap[i] = temp;
		Max_Heapify(Heap, Largest, Len);
	}

}

void Build_Max_Heap(int Heap[], int N){

	for(int i=(N/2)-1; i>=0; i--)
		Max_Heapify(Heap, i, N);

}

void Print(int Heap[], int N){
	for(int i=0; i<N; i++) cout<<Heap[i] <<", ";
}

void Heap_Sort(int Heap[], int N){

	int temp;

	Build_Max_Heap(Heap, N);		//Build Heap
	for(int i=0; i<N; i++){
		temp = Heap[0];		//Swap Root Of Heap with the last element
		Heap[0] = Heap[N-i-1];
		Heap[N-i-1] = temp;
		Max_Heapify(Heap, 0, N-i-1);  //Heapify again from root, reduce length to ignore the last (sorted) elements in array
	}
}

int main(){

	int Heap[] = {13,-3,-25,20,-3,-16,-23,18,20,-7,12,-5,-22,15,-4,7}, N=16;
	cout<<"Orignal array = ";
	Print(Heap, N);		//Print Orignal Array
	Heap_Sort(Heap, N);		//Heap Sort
	cout<<"\n Sorted array = ";
	Print(Heap, N);		//Print Sorted Array
	cout<<endl;
}
